### ISSUE TITLE
LDK gateway panic due to `block_on` in LDK

### ISSUE THREAD

**Issue by dpc** [fedimint/fedimint#6159]
```
thread 'tokio-runtime-worker' panicked at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/mod.rs:86:9:
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.
stack backtrace:
[2m2024-10-11T22:18:32.452969Z[0m [32m INFO[0m [2mgatewayd[0m[2m:[0m Gatewayd exiting...
[2m2024-10-11T22:18:32.453000Z[0m [32m INFO[0m [2mln_gateway::rpc::rpc_server[0m[2m:[0m Successfully shutdown webserver
   0: rust_begin_unwind
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:74:14
   2: tokio::runtime::context::runtime::enter_runtime
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context/runtime.rs:68:5
   3: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
   4: tokio::runtime::runtime::Runtime::block_on_inner
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/runtime.rs:363:45
   5: tokio::runtime::runtime::Runtime::block_on
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/runtime.rs:335:13
   6: ldk_node::Node::stop
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ldk-node-0.3.0/src/lib.rs:827:21
   7: <ln_gateway::lightning::ldk::GatewayLdkClient as core::ops::drop::Drop>::drop
             at ./gateway/ln-gateway/src/lightning/ldk.rs:188:25
   8: core::ptr::drop_in_place<ln_gateway::lightning::ldk::GatewayLdkClient>
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:1
   9: core::ptr::drop_in_place<dyn ln_gateway::lightning::ILnRpcClient>
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:1
  10: alloc::sync::Arc<T,A>::drop_slow
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/sync.rs:1844:18
  11: <alloc::sync::Arc<T,A> as core::ops::drop::Drop>::drop
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/sync.rs:2531:13
  12: core::ptr::drop_in_place<alloc::sync::Arc<dyn ln_gateway::lightning::ILnRpcClient>>
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:1
  13: core::ptr::drop_in_place<ln_gateway::lightning::LightningContext>
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:1
  14: core::ptr::drop_in_place<ln_gateway::GatewayState>
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:1
  15: ln_gateway::Gateway::set_gateway_state::{{closure}}
             at ./gateway/ln-gateway/src/lib.rs:753:9
  16: ln_gateway::Gateway::start_gateway::{{closure}}::{{closure}}
             at ./gateway/ln-gateway/src/lib.rs:462:77
  17: fedimint_core::task::TaskGroup::spawn::{{closure}}
             at ./fedimint-core/src/task.rs:153:35
  18: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.40/src/instrument.rs:321:9
  19: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/future/future.rs:123:9
  20: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:331:17
  21: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/loom/std/unsafe_cell.rs:16:9
  22: tokio::runtime::task::core::Core<T,S>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:320:13
  23: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:500:19
  24: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panic/unwind_safe.rs:272:9
  25: std::panicking::try::do_call
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:557:40
  26: __rust_try
  27: std::panicking::try
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:521:19
  28: std::panic::catch_unwind
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panic.rs:350:14
  29: tokio::runtime::task::harness::poll_future
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:488:18
  30: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:209:27
  31: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:154:15
  32: tokio::runtime::task::raw::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:271:5
  33: tokio::runtime::task::raw::RawTask::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:201:18
  34: tokio::runtime::task::LocalNotified<S>::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/mod.rs:436:9
  35: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:661:17
  36: tokio::runtime::coop::with_budget
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/coop.rs:107:5
  37: tokio::runtime::coop::budget
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/coop.rs:73:5
  38: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:597:9
  39: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:548:24
  40: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:513:21
  41: tokio::runtime::context::scoped::Scoped<T>::set
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context/scoped.rs:40:9
  42: tokio::runtime::context::set_scheduler::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context.rs:180:26
  43: std::thread::local::LocalKey<T>::try_with
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/thread/local.rs:283:12
  44: std::thread::local::LocalKey<T>::with
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/thread/local.rs:260:9
  45: tokio::runtime::context::set_scheduler
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context.rs:180:9
  46: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:508:9
  47: tokio::runtime::context::runtime::enter_runtime
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/context/runtime.rs:65:16
  48: tokio::runtime::scheduler::multi_thread::worker::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:500:5
  49: tokio::runtime::scheduler::multi_thread::worker::block_in_place::{{closure}}::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/scheduler/multi_thread/worker.rs:441:41
  50: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/task.rs:42:21
  51: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:331:17
  52: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/loom/std/unsafe_cell.rs:16:9
  53: tokio::runtime::task::core::Core<T,S>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/core.rs:320:13
  54: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:500:19
  55: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panic/unwind_safe.rs:272:9
  56: std::panicking::try::do_call
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:557:40
  57: __rust_try
  58: std::panicking::try
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:521:19
  59: std::panic::catch_unwind
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panic.rs:350:14
  60: tokio::runtime::task::harness::poll_future
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:488:18
  61: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:209:27
  62: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/harness.rs:154:15
  63: tokio::runtime::task::raw::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:271:5
  64: tokio::runtime::task::raw::RawTask::poll
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/raw.rs:201:18
  65: tokio::runtime::task::UnownedTask<S>::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/task/mod.rs:473:9
  66: tokio::runtime::blocking::pool::Task::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:160:9
  67: tokio::runtime::blocking::pool::Inner::run
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:518:17
  68: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/dpc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.40.0/src/runtime/blocking/pool.rs:476:13
n
```

**Comment by tvolk131** [comment]
This appears to be an upstream issue in `ldk-node`. See https://github.com/lightningdevkit/ldk-node/issues/372

**Comment by vincenzopalazzo** [comment]
Probably fixed by https://github.com/fedimint/fedimint/pull/6172 ?

**Comment by Kodylow** [comment]
@m1sterc001guy there's a start_with_runtime we need to use

**Comment by vincenzopalazzo** [comment]
>@m1sterc001guy there's a start_with_runtime we need to use

I am looking at this part of the code RN

**Comment by vincenzopalazzo** [comment]
?there's a start_with_runtime we need to use

This looks not trivial because with tokio it is possible to get the current runtime with the `Handler` interface and not with the `Runtime` one. So if we would like to pass down the `Runtime` to the `LightningGateway` we need to take the runtime from the `main` and pass down till we reach the ldk-node API.

However, wondering with @m1sterc001guy and @tvolk131 have a better idea with it?

**Comment by tvolk131** [comment]
> we need to take the runtime from the `main` and pass down till we reach the ldk-node API

Yep, not ideal... I opened a PR that does this: #6229

**Comment by tvolk131** [comment]
I dug into the tokio documentation for [`tokio::runtime::Runtime`](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html), and it states the following:

<img width="952" alt="Screenshot 2024-10-26 at 1 35 18 PM" src="https://github.com/user-attachments/assets/26613b84-9756-484b-bbc5-e1e67f91f2f3">

> The difference between [Arc](https://doc.rust-lang.org/nightly/alloc/sync/struct.Arc.html)<Runtime> and [Handle](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html) is that an [Arc](https://doc.rust-lang.org/nightly/alloc/sync/struct.Arc.html)<Runtime> will prevent the runtime from shutting down, whereas a [Handle](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html) does not prevent that. This is because shutdown of the runtime happens when the destructor of the Runtime object runs.

While it'd be easier if the `ldk-node::Node` API relied on a handle, I think a runtime makes more sense since its shutdown command calls async code.

**Comment by vincenzopalazzo** [comment]
>While it'd be easier if the ldk-node::Node API relied on a handle, I think a runtime makes more sense since its shutdown command calls async code.

Yes for shutdown make sense to me, thanks for providing the explaination

**Comment by tvolk131** [comment]
Fixed by #6229
